File: clients/swift/Sources/Readiness/Features/SubinvokePluginWrap.swift
import PolywrapClient
import Foundation


public class SubinvokePlugin: PluginModule {
    public var uri: Uri
    public var method: String
    public var args: [UInt8]

    public init(_ uri: Uri, _ method: String, _ args: [UInt8]) {
        self.uri = uri
        self.method = method
        self.args = args
    }

    public func performSubinvoke(args: EmptyArgs, _: EmptyEnv?, invoker: Invoker) throws -> Bool {
        let result = try invoker.invoke(uri: self.uri, method: self.method, args: self.args, env: nil, resolution_context: nil)
        return true
    }
}

struct SubinvokePluginWrapTest: Feature {
    func runTestCase(input: Any) throws -> Void {
        guard let inputObj = input as? [AnyHashable: Any] else {
           throw InputError.expectedObject
        }

        let wrapDir = try Input.expectRootDir(inputObj["directory"]!, rootDir: "./")
        let method = try Input.expectString(inputObj["method"]!)
        
        guard let argsDict = inputObj["args"] as? [AnyHashable: Any],
            let first = argsDict["first"] as? Int,
            let second = argsDict["second"] as? Int else {
            throw InputError.expectedObject
        }

        let args = u8MethodArgs(first: first, second: second)
        let encoded_args = try encode(value: args)
        
        let pluginUri = try Uri("plugin/bar")
        let wrapperUri = try Uri("embed/foo")

        let subinvokePlugin = SubinvokePlugin(wrapperUri, method, encoded_args)
        subinvokePlugin.addMethod(name: "performSubinvoke", closure: subinvokePlugin.performSubinvoke)

        let pluginPackage = PluginPackage(subinvokePlugin)

        let reader = ResourceReader(bundle: Bundle.module)
        let module = try reader.readFile(wrapDir)
        let wasmPackage = WasmPackage(reader: nil, module: module)
        
        let builder = BuilderConfig().addPackage(wrapperUri, wasmPackage).addPackage(pluginUri, pluginPackage)

        let client = builder.build()
        
        print("Invoking plugin")
        let result: Bool? = try? client.invoke(uri: pluginUri, method: "performSubinvoke")

        guard let r = result else {
            return
        }
        
        print("Received \(r)")
        print("Success")
    }
}

File: clients/py/src/features/subinvoke_plugin_wrap.py
from pathlib import Path
from typing import Any, TypedDict

from polywrap import (
    PluginModule,
    PluginPackage,
    PolywrapClient,
    PolywrapClientConfigBuilder,
    Uri,
    WasmPackage,
)
from pydantic import BaseModel, validator

from validators import validate_root_directory


class Args(TypedDict):
    first: int
    second: int


class TestCaseInput(BaseModel):
    directory: Path
    method: str
    args: Args

    @validator("directory")
    def valid_root_directory(cls, v: Any):
        return validate_root_directory(
            v, Path(__file__).parent.parent.parent.parent.parent
        )


class PerformSubinvokeArgs(TypedDict):
    args: Args
    method: str


class Plugin(PluginModule[None]):
    def __init__(self, config: None):
        super().__init__(config)

    def perform_subinvoke(
        self, args: PerformSubinvokeArgs, client: PolywrapClient, *_: Any
    ):
        print(f"Subinvoking {args['method']}")

        try:
            result = client.invoke(
                uri=Uri.from_str("embed/foo"), method=args["method"], args=args["args"]
            )
            print(f"Received: {result}")
            return True
        except Exception as e:
            print(f"Failed to subinvoke method. {e}")
            return False


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    wrap_dir = Path(__file__).parent / input_obj.directory

    manifest_path = wrap_dir / "wrap.info"
    wasm_path = wrap_dir / "wrap.wasm"

    with open(manifest_path, "rb") as f:
        manifest = f.read()

    with open(wasm_path, "rb") as f:
        wasm_module = f.read()

    wrap_package = WasmPackage(
        file_reader=NotImplemented, manifest=manifest, wasm_module=wasm_module
    )

    plugin_package = PluginPackage(Plugin(None), NotImplemented)

    config = (
        PolywrapClientConfigBuilder()
        .set_package(Uri.from_str("embed/foo"), wrap_package)
        .set_package(Uri.from_str("plugin/bar"), plugin_package)
        .build()
    )

    client = PolywrapClient(config)

    print("Invoking Plugin")

    result = client.invoke(
        uri=Uri.from_str("plugin/bar"),
        method="perform_subinvoke",
        args={
            "method": input_obj.method,
            "args": input_obj.args,
        },
    )

    if result:
        print("Success!")
